### PR TITLE
Mandd/mc ssolver td

### LINKED ISSUE #8: Extend MCS solver to time dependent calculation
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
It is needed to Extend MCS solver to time dependent calculation where temporal profile of the basic events is passed to the MCSsolver. This temporal profile is Boolean in nature: 0 (Basic event set to False: probability=p(BE)) or 1 (Basic event set to True: probability=1.0)

**Describe the solution you'd like**
Extend the MCSsolver to perform actual calculation for each time instant

**Describe alternatives you've considered**

**Additional context**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by mandd**
close #8

### CODE REVIEW

File: doc/user_manual/include/MCSsolver.tex [hunk 1]
```diff
@@ -52,25 +52,94 @@ \section{MCSSolver}
...
+or 1 (Basic event set to True: probability=1.0).
+The format of the dataObject can be:
+\begin{itemize}
+  \item HistorySet: it contains the temporal profile of each basic event (e.g., BE1, BE2, BE3, and BE4) as a time 
+                    series (see test\_MCSSolver_TD.xml). In this case it is needed to specify the ID of the time variable 
+                    contained in the HistorySet in the \xmlNode{timeID} node.
+  
+   \begin{lstlisting}[style=XML,morekeywords={anAttribute},caption=Time dependent (from HistorySet) MCSSolver model input example., label=lst:MCSSolver_InputExample]
+     <Models>
+      <ExternalModel name="MCSmodel" subType="PRAplugin.MCSSolver">
```
Comment (wangcj05) [inline doc/user_manual/include/MCSsolver.tex:89]: change PRAplugin.MCSSolver to SR2ML.MCSSolver

File: doc/user_manual/include/MCSsolver.tex [hunk 2]
```diff
@@ -52,25 +52,94 @@ \section{MCSSolver}
...
+        <map var='statusBE1'>BE1</map>
+        <map var='statusBE2'>BE2</map>
+        <map var='statusBE3'>BE3</map>
+        <map var='statusBE4'>BE4</map>
+      </ExternalModel>
+    </Models>
+  \end{lstlisting}
+  
+  \item PointSet: it contains the interval time (initial and final time) under which each basic event is set to 
+                  True (see test\_MCSSolver_TD_fromPS.xml). As an example, the PointSet contains the IDs of the basic events
```
Comment (wangcj05) [inline doc/user_manual/include/MCSsolver.tex:103]: test\_MCSSolver_TD_fromPS.xml --> test\_MCSSolver\_TD\_fromPS.xml

File: doc/user_manual/include/MCSsolver.tex [hunk 3]
```diff
@@ -52,25 +52,94 @@ \section{MCSSolver}
...
+                  
+   \begin{lstlisting}[style=XML,morekeywords={anAttribute},caption= Example of PointSet for time dependent MCSSolver calculation., label=lst:MCSSolver_InputExample]
+    <PointSet name="maintenanceSchedule_PointSet">
+      <Input>BE</Input>
+      <Output>tIn,tFin</Output>
+    </PointSet>
+    \end{lstlisting}
+    
+                  In this case, the MCSSolver requires the specification of the DataObject variables that contain the list of basic events
+                  (\xmlNode{BE_ID} node) and the initial and final time values (\xmlNode{tInitial} and \xmlNode{tEnd} nodes).
```
Comment (wangcj05) [inline doc/user_manual/include/MCSsolver.tex:114]: BE_ID --> BE\_ID?

File: doc/user_manual/include/MCSsolver.tex [hunk 4]
```diff
@@ -52,25 +52,94 @@ \section{MCSSolver}
...
+                  
+   \begin{lstlisting}[style=XML,morekeywords={anAttribute},caption=Time dependent (from PointSet) MCSSolver model input example., label=lst:MCSSolver_InputExample]
+    <Models>
+      <ExternalModel name="MCSmodel" subType="PRAplugin.MCSSolver">
```
Comment (wangcj05) [inline doc/user_manual/include/MCSsolver.tex:118]: change PRAplugin.MCSSolver to SR2ML.MCSSolver

File: doc/user_manual/include/MCSsolver.tex [hunk 5]
```diff
@@ -52,25 +52,94 @@ \section{MCSSolver}
...
+  \end{lstlisting}
+\end{itemize}
+
 \subsection{MCSSolver model reference tests}
 The following is the provided analytic test:
 \begin{itemize}
-	\item test\_MCSSolver.xml
+  \item test\_MCSSolver.xml
+  \item test\_MCSSolver_TD.xml
+  \item test\_MCSSolver_TD_fromPS.xml
```
Comment (wangcj05) [inline doc/user_manual/include/MCSsolver.tex:140]: test\_MCSSolver_TD.xml --> test\_MCSSolver\_TD.xml
\item test\_MCSSolver\_TD\_fromPS.xml

File: src/MCSSolver.py [hunk 6]
```diff
@@ -39,31 +41,45 @@ def __init__(self):
       @ Out, None
     """
     ExternalModelPluginBase.__init__(self)
+    self.TimeDependentMode= True
```
Comment (wangcj05) [inline src/MCSSolver.py:44]: self.TimeDependentMode --> self.timeDependentMode
Also, please provide inline comments for this variable.

File: src/MCSSolver.py [hunk 7]
```diff
@@ -88,10 +104,18 @@ def createNewInput(self, container, inputs, samplerType, **kwargs):
            a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
       @ Out, kwargs, dict, dictionary which contains the information coming from the sampler
     """
-    if len(inputs) > 1:
+    if len(inputs) > 2:
       raise IOError("MCSSolver: More than one file has been passed to the MCS solver")
-
-    mcsIDs, probability, mcsList, beList = mcsReader(inputs[0])
+    
+    self.timeDepData = None
```
Comment (wangcj05) [inline src/MCSSolver.py:110]: move this to __init__ method

File: src/MCSSolver.py [hunk 8]
```diff
@@ -88,10 +104,18 @@ def createNewInput(self, container, inputs, samplerType, **kwargs):
...
+    for input in inputs:
+      if input.type == 'HistorySet':
+        self.timeDepData = input.asDataset()
+      elif input.type == 'PointSet':
+        self.timeDepData = self.generateHistorySetFromSchedule(container,input.asDataset())
```
Comment (wangcj05) [inline src/MCSSolver.py:115]: predefine self variable in __init__ method

File: src/MCSSolver.py [hunk 9]
```diff
@@ -115,18 +139,70 @@ def createNewInput(self, container, inputs, samplerType, **kwargs):
...
+      teProbability[index] = self.McsSolver(inputForSolver)
+      
+    if container.tdFromPS:
+      for key in container.invMapping.keys():
+        container.__dict__[key] = self.timeDepData[key][0].values
+    
+    container.__dict__[container.timeID]     = self.timeDepData[container.timeID].values
+    container.__dict__[container.topEventID] = teProbability
+    
+  def McsSolver(self, inputDict):
```
Comment (wangcj05) [inline src/MCSSolver.py:199]: McsSolver --> mcsSolver

File: tests/MCSSolverTDfromPS/maintenanceSchedule.xml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<DataObjectMetadata name="sim_PS">
```
Comment (wangcj05) [inline tests/MCSSolverTDfromPS/maintenanceSchedule.xml:1]: do you need this file?

File: tests/test_MCSSolver_TD.xml [hunk 11]
```diff
@@ -0,0 +1,103 @@
+<Simulation verbosity="debug">
+  <!--
+  This is a test for the system: TopEvent = BE1 or BE3 or (BE2 or BE4)
+  MCS = BE1, BE3, BE2 BE4
+  Hence: P(TopEvent) = P(BE1) + P(BE3) + P(BE2)*P(BE4) 
+                     - P(BE1)*P(BE3) - P(BE1)*P(BE2)*P(BE4) - P(BE2)*P(BE3)*P(BE4)
+                     + P(BE1)*P(BE2)*P(BE3)*P(BE4)
+  -->
+  <TestInfo>
+    <name>PRAplugin/MCSSolverTD</name>
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD.xml:10]: change PRAplugin to SR2ML

File: tests/test_MCSSolver_TD.xml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    <author>mandd</author>
+    <created>2020-12-20</created>
+    <classesTested>PRAplugin/MCSSolver</classesTested>
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD.xml:13]: change PRAplugin to SR2ML

File: tests/test_MCSSolver_TD.xml [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    <batchSize>1</batchSize>
+  </RunInfo>
+
+  <Files>
+    <Input name="MCSlistFile"         type="MCSlist"             >MCSlist.csv</Input>
+    <Input name="maintenanceSchedule" type="maintenanceSchedule" >maintenanceSchedule.csv</Input>
+  </Files>
+
+  <Models>
+    <ExternalModel name="MCSmodel" subType="PRAplugin.MCSSolver">
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD.xml:38]: change PRAplugin to SR2ML

File: tests/test_MCSSolver_TD_fromPS.xml [hunk 14]
```diff
@@ -0,0 +1,105 @@
+<Simulation verbosity="debug">
+  <!--
+  This is a test for the system: TopEvent = BE1 or BE3 or (BE2 or BE4)
+  MCS = BE1, BE3, BE2 BE4
+  Hence: P(TopEvent) = P(BE1) + P(BE3) + P(BE2)*P(BE4) 
+                     - P(BE1)*P(BE3) - P(BE1)*P(BE2)*P(BE4) - P(BE2)*P(BE3)*P(BE4)
+                     + P(BE1)*P(BE2)*P(BE3)*P(BE4)
+  -->
+  <TestInfo>
+    <name>PRAplugin/MCSSolverTD_PS</name>
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD_fromPS.xml:10]: change PRAplugin to SR2ML

File: tests/test_MCSSolver_TD_fromPS.xml [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    <author>mandd</author>
+    <created>2021-01-06</created>
+    <classesTested>PRAplugin/MCSSolver</classesTested>
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD_fromPS.xml:13]: change PRAplugin to SR2ML

File: tests/test_MCSSolver_TD_fromPS.xml [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    <batchSize>1</batchSize>
+  </RunInfo>
+
+  <Files>
+    <Input name="MCSlistFile"         type="MCSlist"             >MCSlist.csv</Input>
+    <Input name="maintenanceSchedule" type="maintenanceSchedule" >maintenanceSchedule.csv</Input>
+  </Files>
+
+  <Models>
+    <ExternalModel name="MCSmodel" subType="PRAplugin.MCSSolver">
```
Comment (wangcj05) [inline tests/test_MCSSolver_TD_fromPS.xml:37]: change PRAplugin to SR2ML

**Review by wangcj05 — CHANGES_REQUESTED** [review]
several comments. After you addressed all comments, please make sure the user manual can be compiled and all SR2ML tests are green on your local machine

**Comment by mandd** [comment]
Comments have been addressed and test have been run on local machine.
Board is green on my MAC (see below)
Screen Shot 2021-01-21 at 3 07 45 PM

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
Have you tried to build the manual? @mandd

**Comment by mandd** [comment]
@wangcj05 : good catch, I found an error that I have fixed. Now the user_manual compiles correctly

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
PR is good to me.

**Pull request merged**